iceoryx_posh/test/moduletests/test_popo_listener_waitset_attachments.cpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_posh/popo/client.hpp"
+#include "iceoryx_posh/popo/listener.hpp"
+#include "iceoryx_posh/popo/server.hpp"
+#include "iceoryx_posh/popo/subscriber.hpp"
+#include "iceoryx_posh/popo/untyped_client.hpp"
+#include "iceoryx_posh/popo/untyped_server.hpp"
+#include "iceoryx_posh/popo/untyped_subscriber.hpp"
+#include "iceoryx_posh/testing/mocks/posh_runtime_mock.hpp"
+
+#include "test.hpp"
+namespace
+{
+using namespace ::testing;
+using namespace ::iox::popo;
+using namespace ::iox::cxx;
+using namespace ::iox::capro;
+using namespace ::iox::mepoo;
+
+constexpr const char RUNTIME_NAME[] = "torben_dallas";
+constexpr const char SERVICE[] = "respect_to_the_man_in_the_icecream_van";
+constexpr const char INSTANCE[] = "Lakierski materialski";
+constexpr const char EVENT[] = "boom boom boomerang";
+
+/// Those tests verify that the destructor of the attachables is correct. When we
+/// use inheritance it is possible that the trigger is a member of the base class
+/// and destroyed in the base class constructor as well.
+/// But the listener and waitset require the child class for cleanup. If the child
+/// class destructor does not call reset on all the triggers the base class will
+/// and the listener and waitset will call a method on the original class which is
+/// already deleted. This causes undefined behavior which the sanitizer will catch.
+///
+/// The following tests should be run for all classes which can be attached to a
+/// listener or a waitset to ensure that trigger.reset() is called in the child destructor
```

Comment:
How is this checked in the tests? I assume this is indirectly checked by calling reset and checking for the object to be detached?

Reply:
@elfenpiff This means tests for the `ServiceDiscovery` are still missing here, correct?

Reply:
@mossmaurice good point. We should also add them here